navigation-material/src/main/java/com/google/accompanist/navigation/material/SheetContentHost.kt

```diff
@@ -0,0 +1,147 @@
+import kotlinx.coroutines.currentCoroutineContext
+import kotlinx.coroutines.ensureActive
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.distinctUntilChanged
+import kotlinx.coroutines.flow.drop
+import kotlinx.coroutines.flow.filter
+import kotlinx.coroutines.launch
+
+/**
+ * Hosts a [BottomSheetNavigator.Destination]'s [NavBackStackEntry] and its
+ * [BottomSheetNavigator.Destination.content] and provides a [onSheetDismissed] callback. It also
+ * shows and hides the [ModalBottomSheetLayout] through the [sheetState] when the sheet content
+ * enters or leaves the composition.
+ *
+ * @param columnHost The [ColumnScope] the sheet content is hosted in, typically the instance
+ * that is provided by [ModalBottomSheetLayout]
+ * @param backStackEntry The [NavBackStackEntry] holding the [BottomSheetNavigator.Destination],
+ * or null if there is no [NavBackStackEntry]
+ * @param sheetState The [ModalBottomSheetState] used to observe and control the sheet visibility
+ * @param onSheetDismissed Callback when the sheet has been dismissed. Typically, you'll want to
+ * pop the back stack here.
+ */
+@OptIn(ExperimentalMaterialApi::class)
+@Composable
+internal fun SheetContentHost(
+    columnHost: ColumnScope,
+    backStackEntry: NavBackStackEntry?,
+    sheetState: ModalBottomSheetState,
+    saveableStateHolder: SaveableStateHolder,
+    onSheetShown: (entry: NavBackStackEntry) -> Unit,
+    onSheetDismissed: (entry: NavBackStackEntry) -> Unit
+) {
+    val scope = rememberCoroutineScope()
+    if (backStackEntry != null) {
+        val currentOnSheetShown by rememberUpdatedState(onSheetShown)
+        val currentOnSheetDismissed by rememberUpdatedState(onSheetDismissed)
+        var hideCalled by remember(backStackEntry) { mutableStateOf(false) }
+        LaunchedEffect(backStackEntry, hideCalled) {
+            val sheetVisibility = snapshotFlow { sheetState.isVisible }
+            sheetVisibility
+                // We are only interested in changes in the sheet's visibility
+                .distinctUntilChanged()
+                // distinctUntilChanged emits the initial value which we don't need
+                .drop(1)
+                // We want to know when the sheet was visible but is not anymore
+                .filter { isVisible -> !isVisible }
+                // Finally, pop the back stack when the sheet has been hidden
+                .collect { if (!hideCalled) currentOnSheetDismissed(backStackEntry) }
+        }
+
+        // Whenever the composable associated with the latestEntry enters the composition, we
+        // want to show the sheet, and hide it when this composable leaves the composition
+        DisposableEffect(backStackEntry) {
+            scope.launch {
+                sheetState.internalShow()
+                currentOnSheetShown(backStackEntry)
+            }
+            onDispose {
+                scope.launch {
+                    hideCalled = true
```

Comment:
So... I hate this. This stems from us having two sources of truth: A sheet can be dismissed by the user (that's what we look at in the above snapshotFlow in the LaunchedEffect) and when the associated `backStackEntry` leaves the composition in which case we programmatically hide. When we programmatically hide the sheet, we don't want that visibility flow to emit because it means that we already popped the entry off the back stack. Any ideas would be appreciated, I'm hoping there's a simple-ish solution that I am missing here!

Reply:
Cool, thanks!